shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs

```diff
@@ -50,24 +52,37 @@ genPParams = pure $ emptyPParams {
                                  }
 
 -- | Generate certificates and also return the associated witnesses and
--- total deposits required.
+-- deposits and refunds required.
 genDCerts
   :: KeyPairs
   -> PParams
   -> DPState
-  -> Gen (Seq DCert, [KeyPair], Coin)
-genDCerts keys pparams dpState = do
+  -> Slot
+  -> Gen (Seq DCert, [KeyPair], Coin, Coin)
+genDCerts keys pparams dpState slotWithTTL = do
   -- TODO @uroboros Generate _multiple_ certs per Tx
   cert <- genDCert keys (_dstate dpState)
   case cert of
     Nothing ->
-      return (Seq.empty, [], Coin 0)
+      return (Seq.empty, [], Coin 0, Coin 0)
     Just (cert_, witKey) -> do
       let certs = [cert_]
           witKeys = [witKey]
+          --
           deposits_ = deposits pparams (_stPools (_pstate dpState)) certs
+          --
+          deRegStakeCreds = filter isDeRegKey certs
+          rewardForCred crt =
+            let (dval, dmin, lambda) = decayKey pparams
+             in keyRefund dval
+                          dmin
+                          lambda
+                          (dpState ^. dstate . stKeys)
+                          slotWithTTL
+                          crt
+          refunds_ = sum (rewardForCred <$> deRegStakeCreds)
 
-      return (Seq.fromList certs, witKeys, deposits_)
+      return (Seq.fromList certs, witKeys, deposits_, refunds_)
```

Comment:
is there an ordering in the generated certificates? The sequence of the certificates is traversed like this (to comply with the formal spec)
```
  case certificates of
    Empty -> do
         ...
    certs_ :|> cert -> do
```
which is the inverse of the ordering in the list

Reply:
Not sure I follow...

Which traversal or ordering is inverted? Right now we only generate 1 `DCert` in a singleton list. Or did you mean the ordering of the traversal of the `refunds_` computation?

Reply:
ok, so `certs` is a singleton list for now?
What I was referring to was that when the list `certs` is traversed, one would start from its leftmost element, whereas the `DELEGS` rule starts from the rightmost element of the sequence. That means that if ordering is important (and there is potentially more than one element in `certs`) we might want to reverse `certs` before making it a sequence.

Reply:
Got it. We currently only ever generate one cert, but plan to change that soon. I’ll make a note about the traversal order. Thanks!